File: client/scripts/migrate-tokens.mjs
/**
 * migrate-tokens.mjs
 * Remplace les hex hardcodés (classes Tailwind arbitraires) par des tokens mono.
 * Usage : node scripts/migrate-tokens.mjs
 */

import { readFileSync, writeFileSync, readdirSync, statSync } from 'fs';
import { join, extname, resolve } from 'path';

// Fichiers à exclure (GL shaders + JS color constants)
const EXCLUDED = new Set([
  'ShaderBackground.tsx',
  'MintMistBackground.tsx',
  'PaperFoldBackground.tsx',
  'colors.ts',
]);

// Mapping : hex (uppercase) → token mono
// Ordre important : du plus long au plus court pour éviter les chevauchements
const REPLACEMENTS = [
  // Dark (light scale — high mono number = light color)
  ['#E5E7EB', 'mono-900'],
  ['#F3F4F6', 'mono-950'],
  ['#F5F5F5', 'mono-975'],
  ['#FAFAFA', 'mono-1000'],
  ['#D4D4D4', 'mono-800'],
  ['#A3A3A3', 'mono-700'],
  ['#737373', 'mono-500'],
  ['#525252', 'mono-400'],
  ['#404040', 'mono-350'],
  ['#262626', 'mono-300'],
  ['#1A1A1A', 'mono-200'],
  ['#111111', 'mono-100'],
  ['#0A0A0A', 'mono-50'],
];

// Préfixes Tailwind valides pour les remplacements
// Pattern : `prefix-[#XXXXXX]` → `prefix-mono-NNN`
// On remplace aussi les variantes `dark:prefix-[#XXXXXX]`, `sm:prefix-[#XXXXXX]`, etc.
function buildRegex(hex) {
  // Escape le # pour regex
  const escapedHex = hex.replace('#', '\\#');
  // Match le pattern -[#XXXXXX] (6 chiffres hex, case-insensitive)
  // On veut remplacer SEULEMENT les occurrences dans des classes Tailwind
  // Pattern : word boundary avant, `[#HEX]` (le hex exact)
  return new RegExp(`\\[${escapedHex}\\]`, 'gi');
}

function processFile(filePath) {
  const original = readFileSync(filePath, 'utf8');
  let content = original;

  for (const [hex, token] of REPLACEMENTS) {
    const regex = buildRegex(hex);
    content = content.replace(regex, token);
  }

  if (content !== original) {
    writeFileSync(filePath, content, 'utf8');
    return true;
  }
  return false;
}

function walkDir(dir, extensions = ['.tsx', '.ts', '.jsx', '.js']) {
  const results = [];
  const items = readdirSync(dir);
  for (const item of items) {
    if (item === 'node_modules') continue;
    const fullPath = join(dir, item);
    const stat = statSync(fullPath);
    if (stat.isDirectory()) {
      results.push(...walkDir(fullPath, extensions));
    } else if (extensions.includes(extname(item))) {
      if (!EXCLUDED.has(item)) {
        results.push(fullPath);
      }
    }
  }
  return results;
}

// Main
const srcDir = resolve('./src');
const files = walkDir(srcDir);

let modifiedCount = 0;
let totalFiles = files.length;

console.log(`Scanning ${totalFiles} files...`);

for (const file of files) {
  const modified = processFile(file);
  if (modified) {
    modifiedCount++;
    console.log(`  UPDATED: ${file.replace(srcDir, 'src')}`);
  }
}

console.log(`\nDone. ${modifiedCount}/${totalFiles} files modified.`);

File: client/scripts/migrate-tokens.py
#!/usr/bin/env python3
"""
migrate-tokens.py
Remplace les hex Tailwind hardcodes par des tokens mono dans client/src/.

Usage : python scripts/migrate-tokens.py [--dry-run]
"""

import os
import re
import sys
from pathlib import Path

# --- Configuration ---

EXCLUDED_FILES = {
    'ShaderBackground.tsx',
    'MintMistBackground.tsx',
    'PaperFoldBackground.tsx',
    'colors.ts',
}

EXTENSIONS = {'.tsx', '.ts', '.jsx', '.js'}

# Mapping hex -> token mono
# Ordre : du plus specifique (plus long) au plus general pour eviter les collisions
# Les hex sont en MAJUSCULES pour la normalisation
REPLACEMENTS = [
    ('#E5E7EB', 'mono-900'),
    ('#F3F4F6', 'mono-950'),
    ('#F5F5F5', 'mono-975'),
    ('#FAFAFA', 'mono-1000'),
    ('#D4D4D4', 'mono-800'),
    ('#A3A3A3', 'mono-700'),
    ('#737373', 'mono-500'),
    ('#525252', 'mono-400'),
    ('#404040', 'mono-350'),
    ('#262626', 'mono-300'),
    ('#1A1A1A', 'mono-200'),
    ('#111111', 'mono-100'),
    ('#0A0A0A', 'mono-50'),
]

def process_file(path: Path, dry_run: bool = False) -> tuple[int, int]:
    """Retourne (replacements_count, occurrences_count)"""
    try:
        content = path.read_text(encoding='utf-8')
    except Exception as e:
        print(f'  ERROR reading {path}: {e}')
        return 0, 0

    original = content
    total_replacements = 0

    for hex_val, token in REPLACEMENTS:
        # Pattern: \[#XXXXXX\] (case-insensitive pour matcher #e5e7eb et #E5E7EB)
        escaped = re.escape(f'[{hex_val}]')
        pattern = re.compile(escaped, re.IGNORECASE)
        new_content, count = pattern.subn(token, content)
        content = new_content
        total_replacements += count

    if content != original:
        if not dry_run:
            path.write_text(content, encoding='utf-8')
        return total_replacements, total_replacements
    return 0, 0


def walk_src(src_dir: Path):
    for root, dirs, files in os.walk(src_dir):
        # Exclure node_modules
        dirs[:] = [d for d in dirs if d != 'node_modules']
        for fname in files:
            if Path(fname).suffix in EXTENSIONS:
                if fname not in EXCLUDED_FILES:
                    yield Path(root) / fname


def main():
    dry_run = '--dry-run' in sys.argv
    src_dir = Path(__file__).parent.parent / 'src'

    if not src_dir.exists():
        print(f'ERROR: src directory not found at {src_dir}')
        sys.exit(1)

    print(f'{"DRY RUN - " if dry_run else ""}Scanning {src_dir}...')
    print()

    files = list(walk_src(src_dir))
    modified = []
    total_replacements = 0

    for fpath in files:
        count, _ = process_file(fpath, dry_run=dry_run)
        if count > 0:
            rel = fpath.relative_to(src_dir)
            modified.append((rel, count))
            total_replacements += count
            action = '[DRY RUN]' if dry_run else 'UPDATED'
            print(f'  {action}: src/{rel} ({count} replacements)')

    print()
    print(f'{"=" * 60}')
    print(f'Files modified : {len(modified)}/{len(files)}')
    print(f'Total replacements : {total_replacements}')
    if dry_run:
        print('(No files were written — dry run mode)')


if __name__ == '__main__':
    main()
